### ISSUE TITLE
Foreign Key to hypertable

### ISSUE THREAD

**Issue by jvujcic** [timescale/timescaledb#498]
Allow creation of foreign keys (that contain timestamp) to hypertables.

**Comment by erimatnor** [comment]
@jvujcic Yes, this is a known issue, but a bit tricky to support since the data in the hypertable is actually split up across a number of subtables (chunks). In theory, you should be able to have a foreign key into a specific chunk.

**Comment by mfreed** [comment]
A little more context:  I discussed this issue with @jvujcic at length in Slack, and he understands limitation of handling all types of foreign keys.

However, I think he is especially interested in FKs to existing UNIQUE keys in a hypertable (e.g., whether timestamp is UNIQUE, or a unique composite key on timestamp and some other partitioning key), which should be more architecturally feasible.

**Comment by njern** [comment]
Is this something you guys are planning to tackle? I just noticed this for one of my use cases where I would like to do what the previous poster mentions; refer to existing UNIQUE keys in the hypertable with a foreign key from another hypertable.

**Comment by jvujcic** [comment]
@erimatnor Any news on this limitation?

**Comment by cercata** [comment]
I'm having the same issue while trying to port my DB to make it compatible with timescale.
I understand the limitation, I wonder if there are some "desing practices" on how to link to a hypertable.

I have a table for alarms, and another for alarm ack, and i would like to link to what alarm an ack is referenced.

Can I do the link after I remove the FK ? I mean, the FK is just for ensuring integrity, isn't it ?

**Comment by mashaalmemon** [comment]
Now that timescaledb is finally available on Azure, we are evaluating it for use with our product. This one is a deal breaker. We want to have data integrity and have the benefits of timescale.

Is dealing with this on the roadmap?

**Comment by Norbo11** [comment]
What I've found strange, is that if you have some tables referencing a timestamp column Y in table X and then call `create_hypertable('X', 'Y')`, the call succeeds without any error messages and the foreign keys remain.

Then if you try to add a new foreign key constraint on a hypertable, it fails.

Is there a bug here?

**Comment by kancsuki** [comment]
Seems like [Postgres 12 will be able](https://commitfest.postgresql.org/22/1877/) to reference partitioned tables by FKs.

Does this help in any way to implement the same features for hypertables?

**Comment by lukekhamilton** [comment]
If so how?

**Comment by jvujcic** [comment]
Here is a link to slack conversation:
https://timescaledb.slack.com/archives/C4GT3N90X/p1572502891281000?thread_ts=1572500608.279900&cid=C4GT3N90X

**Comment by nestormh** [comment]
> Here is a link to slack conversation: https://timescaledb.slack.com/archives/C4GT3N90X/p1572502891281000?thread_ts=1572500608.279900&cid=C4GT3N90X

Hi, I tried that example in the current latest release (v2.5.0 over Postgres14 using the docker image timescale/timescaledb:2.5.0-pg14), and it is not working:

```
demo=# create table logs ( evtime timestamp without time zone not null, device_id int not null, something int not null, primary key(evtime, device_id));
CREATE TABLE
demo=# select create_hypertable('logs', 'evtime');
 create_hypertable 
-------------------
 (1,public,logs,t)
(1 row)

demo=# create table logs_extra (  id bigserial,  evtime timestamp without time zone not null, device_id int not null, "data" jsonb not null , FOREIGN KEY (evtime,device_id) REFERENCES logs(evtime, device_id) ON DELETE RESTRICT );
ERROR:  foreign keys to hypertables are not supported
```
Version checking, for reference:
```
demo=# select extname, extversion from pg_extension where extname='timescaledb';
   extname   | extversion 
-------------+------------
 timescaledb | 2.5.0
(1 row)

demo=# select version();
                                                   version                                                    
--------------------------------------------------------------------------------------------------------------
 PostgreSQL 14.0 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20210424) 10.3.1 20210424, 64-bit
(1 row)
```

- I know it is possible to use foreign keys to reference partitioned tables from other tables in Postgres. What is the technical limitation that prevents getting the same behavior in TimescaleDB? I can see that this is now tagged as tech-debt, but I'm curious about the reason.
- Is this issue planned to be addressed in the short term?

Thank you in advance.

**Comment by jvanns** [comment]
I just wanted to add a little +1 to this feature request. I've been able to create two hypertables, one with an fkey to the other. This appears to succeed at schema-creation time. But then an insert fails with this message;

"foreign keys to hypertables are not supported"

So I guess they're not really permitted despite the seeming success of the initial schema creation. The above error is thrown from the 'hidden' query that performs an 'ALTER TABLE _timescaledb_internal._hyper_1_1_chunk ADD CONSTRAINT ....'

**Comment by jomatt** [comment]
+1 on this

I'm a big fan of TimescaleDB so thanks a lot to the whole team for creating such a great product. Is there any comment from the team about this feature request? I (and probably everyone else in this thread) would appreciate a short update.

**Comment by jvanns** [comment]
Awesome! Just to reiterate what @jomatt said - I'm a massive fan of TimescaleDB too and it is such a welcome and needed enhancement to PostgreSQL so the idea of hypertable generalisation sounds fantastic!

**Comment by adriangb** [comment]
I ran into this and worked around it by creating a custom type instead of a table linked by a FK. It works, but custom types can be a pain to work with.

**Comment by Huluti** [comment]
does this is fixed by #6989 and timescaledb 2.16.0?

**Comment by LukasJerabek** [comment]
Thanks for this feature! However, retention policy stops working in case you reference rows that needs to be deleted. Is "delete cascade" option addition to retention planned already or am I the first one to notice?

here in my example rows in table celery.task_destination_map reference celery.events table rows with celery_task_destination_map_fkey foreign key.

```
2024-11-14 12:38:52.181 CET [513510] LOG:  job 1131 threw an error
2024-11-14 12:38:52.182 CET [513510] ERROR:  cannot drop constraint 92_994_events_pkey on table _timescaledb_internal._hyper_1_92_chunk because other objects depend on it
2024-11-14 12:38:52.182 CET [513510] DETAIL:  constraint celery_task_destination_map_fkey on table celery.task_destination_map depends on index _timescaledb_internal."92_994_events_pkey"
2024-11-14 12:38:52.182 CET [513510] HINT:  Use DROP ... to drop the dependent objects.
2024-11-14 12:38:52.185 CET [278411] LOG:  background worker "Retention Policy [1131]" (PID 513510) exited with exit code 1
```

**Comment by akuzm** [comment]
Implemented since 2.16 https://github.com/timescale/timescaledb/pull/6989
